Bin controller

RyanBerge opened the pull request:
This pull requests introduces an intermediary interface between our primary systems and the hardware layer for controlling the tilt of the bin.  It is implemented as an Action Server with a name of `bin_control`.  I can be called with a goal to either raise or lower the bin, and will return when the bin has reached the desired position.

This pull request does not introduce a client component for testing, so this code is mostly untested.

Jebbs committed:
Create tfr_mission_control package

Additionally, this adds the message structure that mission control will receive.

This also removes the tfr_communication package, which was superseded by
the new mission control package.

RyanBerge committed:
BinController: Add a BinController component

RyanBerge committed:
BinController: Refactor BinController component

- Add it to the Controller node, rather than being its own node
- Add placeholder interface for getting signals from the hardware_interface
  layer

collinsc commented:
@RyanBerge  Can you provide a brief pull request description telling me what you did and how you want me to test your code?

collinsc commented on src/tfr_control/src/bin_control_server.cpp:
```diff
@@ -0,0 +1,82 @@
+// I could also pass in the name of the server, but it would end up being a literal
+// string anyway unless it also gets passed into the node, which seems a bit excessive.
+// If it's going to be a literal string we might as well put it as close to its destination
+// as possible so it's clear what it's being used for.
```
Literal strings are fine for things that shouldn't change, or we don't ever really see changing.

collinsc commented on src/tfr_control/src/bin_control_server.cpp:
```diff
@@ -0,0 +1,82 @@
+{
+    std_msgs::Float64 command;
+    tfr_msgs::BinResult result;
+
```
There needs to be a preemption check at the start of every action server execute cb.

collinsc replied:
even if we don't support it during the blocking call.

RyanBerge replied:
What's the reason for this?  I can't imagine a situation where we would immediately preempt an action after requesting it.  Is the preemption flag also used to detect weird internal ROS failures?

collinsc replied:
It's just a best practice, I haven't seen a tutorial without it. It's just one of those things we should automatically do.

collinsc commented on src/tfr_control/src/bin_control_server.cpp:
```diff
@@ -0,0 +1,82 @@
+        result.return_code = result.BIN_LOWERED;
+    }
+
+    server.setSucceeded(result);
```
Is there any reason this could fail? If so how could it fail? Would you ever need to use `setAborted`?

RyanBerge replied:
In this current implementation it uses `setAborted` only if the goal doesn't use one of the predefined inputs (`RAISE_BIN` or `LOWER_BIN`).

Conceivably, the raising of the bin _could_ fail (though I really hope this wouldn't happen), but we don't have a mechanism in place yet to detect that and I didn't write the signaling interface in such a way that it can tell us that, either.  I figured that would be up to a future edit.

collinsc replied:
That was more of my question. If there is no way to detect failure then we don't have to check, but if the provided interface provides a way to check for failure we should do it here.

collinsc commented on src/tfr_control/include/tfr_control/bin_control_server.h:
```diff
@@ -0,0 +1,39 @@
+#include <condition_variable>
+
+using Server = actionlib::SimpleActionServer<tfr_msgs::BinAction>;
+
```
Give me some headers describing what going on in this serve, also describe explictly what it publishes and where in the header.

RyanBerge replied:
Sure.

collinsc commented on src/tfr_control/src/controller.cpp:
```diff
@@ -30,20 +30,33 @@ namespace tfr_control
+    //       If other values read end up needing to be returned, then this should 
+    //       probably return some data structure. It may be that read values get published
+    //       out on topics instead.
+    bool Controller::read() 
```
Let's make this method name a little more specific. If this is supposed to be the controller for all of the hardware interfaces on the whole robot, then methods just like `read` or `write` are not descriptive enough. Maybe `read_bin` or `write_bin` might improve maintainability.

collinsc replied:
Follow up question is `controller.cpp` the only controller class for the whole robot?

RyanBerge replied:
`controller.cpp` is the class we use as an interface between the hardware and our message-passing system, though that interface is also shared by a ROS class called `controller_manager` (both of these classes, as well as my bin control server, live in the `controller_launcher` node).  So yes; `controller.cpp` is for the entire robot.  `read()` reads from our encoders (all of them) and does some as-yet-undefined action with the data.  `write()` writes voltage values out to the I2C connection for all our actuators. 

This was all set up by @AdamuKaapan, so he can correct me if I've gotten anything wrong.

AdamuKaapan replied:
Yeah, that description looks correct. I think that we might want to find a different way to pass out the return value, whether we save them in some other local variable, put them out on topics, or something else @RyanBerge . While I can't *think* of anything else that might need a return value from this method right now, it still feels like a kinda kludgey design and would prefer to find a better solution. And @collinsc due to all of the sensor inputs coming from the same hardware device it makes it much simpler/resource efficient to put it in one method, as much as I want to split it apart (although we could perform separate processing of said inputs on a separate input).

AdamuKaapan replied:
Although now that I've seen the rest and seen that the bin control server is integrated with the controller_launcher itself, this isn't quiiiite as bad. I think I'd still prefer a slightly better solution though.

collinsc commented on src/tfr_control/src/bin_control_server.cpp:
```diff
@@ -0,0 +1,82 @@
+    server.setSucceeded(result);
+}
+
+void BinControlServer::SignalBinController()
```
This synchronization callback needs a comment header describing who calls it and why.

RyanBerge replied:
Sure.

collinsc commented on src/tfr_control/src/controller.cpp:
```diff
@@ -30,20 +30,33 @@ namespace tfr_control
     }
 
-    void Controller::write() {
+    void Controller::write() 
```
see above comment

collinsc commented on src/tfr_control/src/bin_control_server.cpp:
```diff
@@ -0,0 +1,82 @@
+        // https://stackoverflow.com/questions/16350473/why-do-i-need-stdcondition-variable
+        // wait() will release the lock. Sometimes the signal will be notified by accident, so
+        // a check on bin_task_completed is required. wait() reaquires the lock after it unblocks.
+        signal.wait(signal_lock);
```
From what I understand from you code, 
1. client sets goal
2.  server receives it and publishes on topic
3.  block until completed.
  - controller.cpp sets condition variable when it finishes. 
4.  return result.

Thats fine, but it took me too long to get there from reading your code. 

You need to take a few steps to make this more readable and maintainable for future readers:
- Extract all the blocking synchronization code into a separate method: e.g. `wait_for_bin()`
  - In that method tell the readers what's happening, who we are blocking on, and what events trigger this.
  - I think at least 10 minutes of reading could be shaved of "reader comprehension time" if we did this, synchronization usually needs a little extra care in documentation if you want to maintain it..
- If we aren't supporting preemption, explain why in the class header, and in the description of the method.

RyanBerge replied:
Sure.  Yeah, you've got it right.

collinsc reviewed (changes requested):
Most of the code looks good, I asked for some readability changes.

collinsc commented:
@RyanBerge I have a change an additional that I want to request, can you refactor the action server to just successfuly exit and set all of the correct messsages. I can integrate the dumping action server with this next sprint if you do.

RyanBerge commented:
@collinsc What do you mean by "just successfully exit"?  Are you saying do placeholder code so it doesn't block for now?

RyanBerge committed:
BinController: Add better code documentation for maintainability

- Set BinController to not block at all for now since we don't have
  a mechanism yet for detecting the bin position (also we don't
  have a bin)
- Minor refactor in BinControlServer to ease readability
- Something weird happened with a merge conflict in tfr_executive
  while rebasing and I ended up with a line in the CMake that was
  trying to build an executable out of source files that no longer
  existed. So I fixed that.

collinsc commented:
@RyanBerge Yep that way I can integrate it with the dumping server. Thanks!

collinsc commented:
Looks good! I approved the updates, thanks for doing them. It would be great if we had a short description in the pull request header for posterity.

AdamuKaapan commented on src/tfr_control/src/controller.cpp:
```diff
@@ -61,7 +76,8 @@ namespace tfr_control
 
-    double Controller::get_update_time() {
+    double Controller::get_update_time() 
+    {
```
:+1: thanks for fixing the curly brace alignment that I screwed up

AdamuKaapan reviewed (changes requested):
Only request is to investigate into possible alternate solutions to having the `read()` method in `controller` returning a bool. If you come back to the conclusion that this is the best solution then that's alright though.

RyanBerge commented:
@AdamuKaapan I'm definitely not convinced this is the best solution.  My plan was to refactor it once I knew what the design in `read()` and `write()` was going to look like (and writing that hardware interface is my next task).  Do you want me to change it now?

collinsc reviewed (approved):
Thanks for adding the description! I am reapproving after the rebase.

AdamuKaapan commented:
@RyanBerge if you have plans to change it then that's totally fine, I'll go ahead and approve it for now.
